internal/pkg/consensus/expected.go

```diff
@@ -241,67 +242,43 @@ func (c *Expected) validateMining(ctx context.Context, st state.Tree, ts block.T
 }
 
 // runMessages applies the messages of all blocks within the input
-// tipset to the input base state.  Messages are applied block by
-// block with blocks sorted by their ticket bytes.  The output state must be
-// flushed after calling to guarantee that the state transitions propagate.
+// tipset to the input base state.  Messages are extracted from tipset
+// blocks sorted by their ticket bytes and run as a single state transition
+// for the entire tipset. The output state must be flushed after calling to
+// guarantee that the state transitions propagate.
 //
-// An error is returned if individual blocks contain messages that do not
-// lead to successful state transitions.  An error is also returned if the node
-// faults while running aggregate state computation.
-func (c *Expected) runMessages(ctx context.Context, st state.Tree, vms vm.StorageMap, ts block.TipSet, blsMessages [][]*types.UnsignedMessage, secpMessages [][]*types.UnsignedMessage, ancestors []block.TipSet) (state.Tree, []*types.MessageReceipt, error) {
-	var cpySt state.Tree
-	var results []*ApplicationResult
-
-	// TODO: don't process messages twice
+// An error is returned if the state root or receipt root of any block
+// does not match that of the parent tipset.  An error is also returned if
+// the node faults while running aggregate state computation.
+func (c *Expected) runMessages(ctx context.Context, st state.Tree, vms vm.StorageMap, ts block.TipSet, blsMessages [][]*types.UnsignedMessage, secpMessages [][]*types.UnsignedMessage, parentReceipts cid.Cid, ancestors []block.TipSet) (state.Tree, []*types.MessageReceipt, error) {
+	parentStateRoot, err := st.Flush(ctx)
```

Comment:
This flush just to get hold of the parentStateRoot is a bit icky. The caller has this parentStateRoot - that's now the tree was loaded in the first place. The name `runMessages` is also a bit misleading if it also does this other validation.

I suggest pulling the state and receipt root match checks up to `RunStateTransition`.

Reply:
Agreed, but it really makes more sense to put it in `validateMining` since that function is already iterating through blocks and checking for mistakes. I'm now passing the state root and receipt cids into that function so we don't need to flush the state tree.